_posts/2019-06-03-rambling-javascript-5-spread.md

```diff
@@ -0,0 +1,294 @@
+
+```
+const array = [ 1, 2, 3 ];
+const arrayPush = [ 4, 5 ];
+
+const result1 = Array.prototype.push.apply(array, arrayPush);
+const result2 = array.push(...arrayPush);
+// result1 & result2 = [ 1, 2, 3, 4, 5 ]
+```
+
+Al expandir los elementos de ```arrayPush``` en la variable ```array``` podemos usar la función _push_ directamente sin tener que recurrir al _push_ de la clase ```Array``` y todo queda mucho más legible.
+
+### Concatenación de _arrays_
+
+Para concatenar dos _arrays_ se usa la función definida en el prototype de ```Array``` _concat_. Podemos obtener el mismo resultado con _concat_ que usando _spread_:
+
+```js
+const array = [ 4, 5, 6 ];
+const result = [ 1, 2, 3 ].concat(array);
+
+const resultWithSpread = [ 1, 2, 3, ...array]
+
+// result & resultWithSpread = [ 1, 2, 3, 4, 5, 6 ]
+```
+
+En este caso, el resultado de concatenar los _arrays_ es el mismo; pero en mi opinión el operador ```...``` le da mucha más expresividad a nuestro código ya que lo hace mucho más fácil de leer.
+
+### Operación _splice_ en un _array_
+
+La función [_splice_](https://developer.mozilla.org/es/docs/Web/JavaScript/Referencia/Objetos_globales/Array/splice) combina dos _arrays_. Es una función algo compleja que recibe como parámetros el inicio de donde se inserta el nuevo _array_, el número de elementos que sustituye y el nuevo _array_:
+
+```js
+const numbers = [ 4, 5 ];
+numbers.splice(0, 0, [1, 2, 3]);
+// numbers: [ 1, 2, 3, 4, 5]
+
+const numbers = [ 1, 2, 2, 4, 5 ];
+numbers.splice(1, 2, [3, 3]);
+// numbers: [ 1, 3, 3, 4, 5]
+
+const numbers = [ 1, 2, 3, 5 ];
+numbers.splice(3, 0, [4]);
+// numbers: [ 1, 2, 3, 4, 5]
+```
+
+Estas operaciones usando el operador _spread_ pueden quedar mucho más legibles. Para ver donde colocar el nuevo _array_ nos ayudaremos de la función [_slice_](https://developer.mozilla.org/es/docs/Web/JavaScript/Referencia/Objetos_globales/Array/slice), que nos da el subconjunto del _array_ que necesitamos:
+
+```js
+const array = [ 1, 2, 3 ];
+const numbers = [ 4, 5 ];
+const result = [
+    ...array,
+    ...numbers
+];
+// result: [ 1, 2, 3, 4, 5]
+numbers.splice(0, 0, array);
+// numbers: [ 1, 2, 3, 4, 5]
+
+const array = [ 3, 3 ];
+const numbers = [ 1, 2, 2, 4, 5 ];
+const result = [
+    ...numbers.slice(0, 1),
+    ...array,
+    ...numbers.slice(3)
+];
+// result: [ 1, 3, 3, 4, 5]
+numbers.splice(1, 2, array);
+// numbers: [ 1, 3, 3, 4, 5]
+
+const array = [ 4 ];
+const numbers = [ 1, 2, 3, 5 ];
+const result = [
+    ...numbers.slice(0, 3),
+    ...array,
+    ...numbers.slice(3)
+];
+// result: [ 1, 2, 3, 4, 5]
+numbers.splice(3, 0, array);
+// numbers: [ 1, 2, 3, 4, 5]
+
+```
+
+Si lo pensamos con detenimiento una vez vista la función _slice_ esto es equivalente:
+
+```
+const array = [ 1, 2, 3 ];
+
+const result1 = [ array.slice() ];
+const result2 = [ ...array ];
+
+// result1 & result2 = [ 1, 2, 3 ]
+
+```
+
+## Operador _spread_ como parámetros de una función
```

Comment:
Sería parámetro rest, no operador spread

Reply:
Diferencia de agrupar elementos a separar elementos individualmente

Reply:
Complete this.